nbgrader/preprocessors/instantiatetests.py

```diff
@@ -0,0 +1,420 @@
+        """
+        attempts to load the tests.yml file within the assignment directory. In case such file is not found
+        or perhaps cannot be loaded, it will attempt to load the default_tests.yaml file with the course_directory
+        """
+        self.log.debug('loading template tests.yml...')
+        self.log.debug(f'kernel_name: {resources["kernel_name"]}')
+        try:
+            with open(os.path.join(resources['metadata']['path'], self.autotest_filename), 'r') as tests_file:
+                tests = yaml.safe_load(tests_file)
+            self.log.debug(tests)
+
+        except FileNotFoundError:
+            # if there is no tests file, just load a default tests dict
+            self.log.warning(
+                'No tests.yml file found in the assignment directory. Loading the default tests.yml file in the course root directory')
+            # tests = {}
+            try:
+                with open(os.path.join(self.autotest_filename), 'r') as tests_file:
+                    tests = yaml.safe_load(tests_file)
+            except FileNotFoundError:
+                # if there is no tests file, just create a default empty tests dict
+                self.log.warning(
+                    'No tests.yml file found. If AUTOTESTS appears in testing cells, an error will be thrown.')
+                tests = {}
+            except yaml.parser.ParserError as e:
+                self.log.error('tests.yml contains invalid YAML code.')
+                self.log.error(e.msg)
+                raise
+
+        except yaml.parser.ParserError as e:
+            self.log.error('tests.yml contains invalid YAML code.')
+            self.log.error(e.msg)
+            raise
+
+        # get kernel specific data
+        tests = tests[resources["kernel_name"]]
+
+        # get the test templates
+        self.test_templates_by_type = tests['templates']
+
+        # get the test dispatch code template
+        self.dispatch_template = tests['dispatch']
+
+        # get the success message template
+        self.success_code = tests['success']
+
+        # get the hash code template
+        self.hash_template = tests['hash']
+
+        # get the hash code template
+        self.check_template = tests['check']
+
+        # get the hash code template
+        self.normalize_template = tests['normalize']
+
+        # get the setup code if it's there
+        self.setup_code = tests.get('setup', None)
+
+    # -------------------------------------------------------------------------------------
+    def _instantiate_tests(self, snippet, salt=None):
+        # get the type of the snippet output (used to dispatch autotest)
+        template = j2.Environment(loader=j2.BaseLoader).from_string(self.dispatch_template)
+        dispatch_code = template.render(snippet=snippet)
+        dispatch_result = self._execute_code_snippet(dispatch_code)
+        self.log.debug('Dispatch result returned by kernel: ', dispatch_result)
```

Comment:
There is a warning about arguments:
```suggestion
        self.log.debug(f'Dispatch result returned by kernel: {dispatch_result}')
```

Reply:
Ah, actually I'll use something like the below for consistency with the rest of the code:

```
./nbgrader/converters/base.py:        self.log.debug("Student: %s", gd['student_id'])
```